File: yarvis-IA/parseador_de_tickets/cerebro/lote.py
from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse
from pydantic import BaseModel
import sqlite3
import json
import os
import glob
import asyncio

from .analizador import MapeoColumnas, _parsear_linea as parsear_linea, _extraer_fecha_hora_regex, _extraer_metodo_pago
from parseador_de_tickets.llm.analizador_llm import descargar_modelos

router = APIRouter()


def _descargar_todos_los_modelos():
    """Libera RAM/VRAM de todos los modelos Qwen (parser y chat) al terminar un parseo."""
    from chatbot.motor_chat.modelos_local.gestion_hardware import descargar_modelo
    from chatbot.motor_chat.modelos_local.variables import MODELOS

    try:
        descargar_modelos()
    except Exception:
        pass
    for key in MODELOS:
        try:
            descargar_modelo(key)
        except Exception:
            pass


class ParseCarpetaRequest(BaseModel):
    carpeta: str
    mapeo: MapeoColumnas
    db_path: str


def _obtener_archivos_txt(carpeta: str) -> list[str]:
    patron = os.path.join(carpeta, "*.txt")
    archivos = glob.glob(patron)
    archivos.sort()
    return archivos


def _calcular_subtotal(items: list[dict]) -> float:
    return sum(i.get("total", 0) or (i.get("cantidad", 0) * i.get("precio_unitario", 0)) for i in items)


def _extraer_cajero(texto: str) -> str:
    for linea in texto.splitlines()[:10]:
        lower = linea.lower()
        if "cajero" in lower or "empleado" in lower or "vendedor" in lower:
            partes = linea.split(":", 1)
            if len(partes) == 2:
                return partes[1].strip()
    return "SISTEMA"


def _insertar_venta(conn: sqlite3.Connection, items: list[dict], cajero: str, fecha_iso: str = None, metodo_pago: str = "efectivo") -> int:
    subtotal = _calcular_subtotal(items)
    iva = round(subtotal * 0.16, 2)
    total = round(subtotal + iva, 2)

    if fecha_iso:
        cursor = conn.execute("""
            INSERT INTO ventas (total, subtotal, iva, cajero, metodo_pago, estado, fecha)
            VALUES (?, ?, ?, ?, ?, ?, ?)
        """, (total, subtotal, iva, cajero, metodo_pago, "completada", fecha_iso))
    else:
        cursor = conn.execute("""
            INSERT INTO ventas (total, subtotal, iva, cajero, metodo_pago, estado)
            VALUES (?, ?, ?, ?, ?, ?)
        """, (total, subtotal, iva, cajero, metodo_pago, "completada"))
    venta_id = cursor.lastrowid

    for item in items:
        cant = item.get("cantidad", 1)
        precio = item.get("precio_unitario", 0)
        desc = item.get("descuento", 0) or 0
        sub = round(cant * precio - desc, 2)

        conn.execute("""
            INSERT INTO detalle_ventas (venta_id, producto_nombre, cantidad, precio_unitario, descuento, subtotal)
            VALUES (?, ?, ?, ?, ?, ?)
        """, (venta_id, item["producto"], cant, precio, desc, sub))

        # Actualizar stock y vendido en productos (case-insensitive)
        conn.execute("""
            UPDATE productos SET stock = stock - ? WHERE LOWER(nombre) = LOWER(?)
        """, (cant, item["producto"]))
        conn.execute("""
            UPDATE productos SET vendido = vendido + ? WHERE LOWER(nombre) = LOWER(?)
        """, (cant, item["producto"]))

    return venta_id


def _cargar_estado(db_path: str) -> dict:
    state = {"productos_db": {}, "duplicados": 0}
    try:
        conn = sqlite3.connect(db_path)
        rows = conn.execute("SELECT id, nombre, precio_venta FROM productos").fetchall()
        for pid, nombre, precio in rows:
            key = f"{nombre.upper().strip()}|{precio:.2f}"
            state["productos_db"][key] = pid
        conn.close()
    except Exception:
        pass
    return state


def _procesar_archivos(archivos: list[str], mapeo: dict, db_path: str):
    """Procesa cada archivo con transacción propia; cede un dict por archivo.

    Devuelve por cada archivo algo así:
        {
          "archivo", "ok", "motivo", "items", "duplicados",
          "nuevos", "existentes", "venta_id", "total",
        }

    Es la ÚNICA implementación del bucle de parseo: tanto el modo síncrono
    (/parsear_carpeta) como el streaming (/parsear_carpeta_stream) lo
    recorren, evitando la duplicación de ~80 líneas que ya había divergido.

    Transaccionalidad: cada archivo abre UNA transacción (BEGIN...COMMIT).
    Si _insertar_venta falla a mitad (ej. producto inexistente), se hace
    rollback: la venta parcial y los UPDATEs de stock de ese archivo se
    descartan. NUNCA queda un insert a medias commiteado.
    """
    state = _cargar_estado(db_path)
    conn = sqlite3.connect(db_path)
    mapeo_obj = MapeoColumnas(**mapeo)

    def _info(ok, motivo, **kw):
        base = {
            "archivo": "", "ok": ok, "motivo": motivo, "items": 0,
            "duplicados": 0, "nuevos": [], "existentes": 0,
            "venta_id": None, "total": 0.0,
        }
        base.update(kw)
        return base

    try:
        for archivo in archivos:
            nombre_archivo = os.path.basename(archivo)
            try:
                with open(archivo, "r", encoding="utf-8", errors="ignore") as f:
                    texto = f.read()

                if not texto.strip():
                    yield _info(False, "archivo vacío", archivo=nombre_archivo)
                    continue

                lineas = [l for l in texto.strip().splitlines() if l.strip()]
                if not lineas:
                    yield _info(False, "sin líneas útiles", archivo=nombre_archivo)
                    continue

                total_cols = max(len(l.split()) for l in lineas)
                items = []
                seen = set()
                duplicados = 0
                existentes = 0
                nuevos = []

                for linea in lineas:
                    try:
                        item = parsear_linea(linea, mapeo_obj, total_cols)
                    except Exception as e:
                        print(f"[YARVIS-PARSER] Error parseando línea en {nombre_archivo}: {e}")
                        continue
                    if not item:
                        continue
                    dup_key = f"{item['producto']}|{item.get('precio_unitario', 0):.2f}"
                    if dup_key in seen:
                        duplicados += 1
                        continue
                    seen.add(dup_key)
                    if dup_key in state["productos_db"]:
                        existentes += 1
                    else:
                        state["productos_db"][dup_key] = None
                        nuevos.append({"nombre": item["producto"], "precio": item.get("precio_unitario", 0)})
                    items.append(item)

                if not items:
                    yield _info(False, "ningún producto reconocido con el mapeo actual", archivo=nombre_archivo)
                    continue

                fecha, hora = _extraer_fecha_hora_regex(texto)
                fecha_iso = None
                if fecha:
                    fecha_iso = f"{fecha} {hora}:00" if hora else f"{fecha} 00:00:00"

                cajero = _extraer_cajero(texto)
                metodo_pago = _extraer_metodo_pago(texto)

                try:
                    conn.execute("BEGIN")
                    try:
                        venta_id = _insertar_venta(conn, items, cajero, fecha_iso, metodo_pago)
                        conn.commit()
                    except Exception:
                        conn.rollback()
                        raise
                except Exception as e:
                    yield _info(
                        False, f"error al insertar en DB: {e}",
                        archivo=nombre_archivo, items=len(items),
                        duplicados=duplicados, nuevos=nuevos, existentes=existentes,
                    )
                    continue

                yield _info(
                    True, None,
                    archivo=nombre_archivo, items=len(items), duplicados=duplicados,
                    nuevos=nuevos, existentes=existentes, venta_id=venta_id,
                    total=round(_calcular_subtotal(items) * 1.16, 2),
                )
            except Exception as e:
                print(f"[YARVIS-PARSER] Error inesperado procesando {nombre_archivo}: {e}")
                yield _info(False, f"error inesperado: {e}", archivo=nombre_archivo)
    finally:
        conn.close()


def _procesar_carpeta_impl(archivos: list[str], mapeo: dict, db_path: str) -> dict:
    stats = {
        "total_archivos": len(archivos),
        "procesados": 0,
        "exitosos": 0,
        "errores": 0,
        "ventas_creadas": 0,
        "items_insertados": 0,
        "productos_nuevos": 0,
        "productos_existentes": 0,
        "duplicados_detectados": 0,
        "productos_nuevos_lista": [],
        "resumen_ventas": [],
        "tickets_fallidos": [],
    }

    nombres_nuevos_vistos: set = set()

    for res in _procesar_archivos(archivos, mapeo, db_path):
        stats["procesados"] += 1
        if res["ok"]:
            stats["exitosos"] += 1
            stats["ventas_creadas"] += 1
            stats["items_insertados"] += res["items"]
            stats["duplicados_detectados"] += res["duplicados"]
            stats["productos_existentes"] += res["existentes"]
            stats["productos_nuevos"] += len(res["nuevos"])
            for nuevo in res["nuevos"]:
                if nuevo["nombre"] not in nombres_nuevos_vistos:
                    nombres_nuevos_vistos.add(nuevo["nombre"])
                    stats["productos_nuevos_lista"].append(nuevo)
            stats["resumen_ventas"].append({
                "archivo": res["archivo"],
                "venta_id": res["venta_id"],
                "items": res["items"],
                "total": res["total"],
            })
        else:
            stats["errores"] += 1
            stats["tickets_fallidos"].append({"archivo": res["archivo"], "motivo": res["motivo"]})

    stats["productos_nuevos_lista"] = stats["productos_nuevos_lista"][:100]
    stats["tickets_fallidos"] = stats["tickets_fallidos"][:500]  # máx 500 para no inflar la respuesta
    return stats


@router.post("/parsear_carpeta")
async def parsear_carpeta(request: ParseCarpetaRequest):
    carpeta = request.carpeta
    mapeo = request.mapeo.model_dump()
    db_path = request.db_path

    if not os.path.isdir(carpeta):
        raise HTTPException(status_code=400, detail=f"Carpeta no encontrada: {carpeta}")

    archivos = _obtener_archivos_txt(carpeta)
    if not archivos:
        raise HTTPException(status_code=400, detail="No se encontraron archivos .txt en la carpeta")

    descargar_modelos()

    try:
        stats = _procesar_carpeta_impl(archivos, mapeo, db_path)
    finally:
        _descargar_todos_los_modelos()

    return {
        "status": "ok",
        **stats,
    }


@router.post("/parsear_carpeta_stream")
async def parsear_carpeta_stream(request: ParseCarpetaRequest):
    carpeta = request.carpeta
    mapeo = request.mapeo.model_dump()
    db_path = request.db_path

    if not os.path.isdir(carpeta):
        raise HTTPException(status_code=400, detail=f"Carpeta no encontrada: {carpeta}")

    archivos = _obtener_archivos_txt(carpeta)
    if not archivos:
        raise HTTPException(status_code=400, detail="No se encontraron archivos .txt en la carpeta")

    total = len(archivos)

    descargar_modelos()

    async def event_generator():
        procesados = 0
        exitosos = 0
        errores = 0
        ventas_creadas = 0
        items_insertados = 0
        productos_nuevos = 0
        productos_existentes = 0
        duplicados_detectados = 0
        productos_nuevos_set = set()
        tickets_fallidos = []  # lista de {archivo, motivo} para los que no se parsearon

        try:
            # Misma lógica de parseo/inserción que el modo síncrono; aquí solo
            # se acumulan contadores y se emiten eventos de progreso.
            for res in _procesar_archivos(archivos, mapeo, db_path):
                procesados += 1
                if res["ok"]:
                    exitosos += 1
                    ventas_creadas += 1
                    items_insertados += res["items"]
                    duplicados_detectados += res["duplicados"]
                    productos_existentes += res["existentes"]
                    for nuevo in res["nuevos"]:
                        productos_nuevos_set.add(nuevo["nombre"])
                    productos_nuevos += len(res["nuevos"])
                else:
                    errores += 1
                    tickets_fallidos.append({"archivo": res["archivo"], "motivo": res["motivo"]})

                if procesados % 50 == 0 or procesados == total:
                    yield f"data: {json.dumps({'type': 'progress', 'procesados': procesados, 'total': total, 'exitosos': exitosos, 'errores': errores, 'ventas_creadas': ventas_creadas, 'items_insertados': items_insertados, 'productos_nuevos': productos_nuevos, 'productos_existentes': productos_existentes, 'duplicados_detectados': duplicados_detectados})}\n\n"
                    await asyncio.sleep(0.01)

            yield f"data: {json.dumps({'type': 'complete', 'total_archivos': total, 'procesados': procesados, 'exitosos': exitosos, 'errores': errores, 'ventas_creadas': ventas_creadas, 'items_insertados': items_insertados, 'productos_nuevos': productos_nuevos, 'productos_existentes': productos_existentes, 'duplicados_detectados': duplicados_detectados, 'productos_nuevos_lista': list(productos_nuevos_set)[:100], 'tickets_fallidos': tickets_fallidos[:500]})}\n\n"

        finally:
            # Siempre se ejecuta: aunque el cliente cierre la conexión SSE a mitad del proceso
            _descargar_todos_los_modelos()

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        }
    )

File: src-ia/src/cerebro/lote.rs
//! lote.rs — Port de `yarvis-IA/parseador_de_tickets/cerebro/lote.py`
//!
//! Procesa carpetas de tickets .txt con transacción propia por archivo.
//! Sin HTTP: expone las funciones puras de parseo/inserción, consumibles
//! desde un servidor axum o los comandos de Tauri (la capa HTTP es aparte).
//!
//! Depende de `cerebro::analizador` (parsear_linea, fecha/hora, pago) que a
//! su vez conecta con `cerebro::filtrador` (limpiar_producto).

use rusqlite::{params, Connection};
use std::collections::HashSet;
use std::path::Path;
use std::sync::mpsc::Sender;

use super::analizador::{extraer_fecha_hora_regex, extraer_metodo_pago, parsear_linea, Item, MapeoColumnas};

// ---------------------------------------------------------------------------
// Tipos
// ---------------------------------------------------------------------------

#[derive(Debug, Clone, serde::Serialize)]
pub struct ProductoNuevo {
    pub nombre: String,
    pub precio: f64,
}

/// Resultado de procesar UN archivo (equivalente a un `yield` de Python).
#[derive(Debug, Clone, serde::Serialize)]
pub struct ArchivoResultado {
    pub archivo: String,
    pub ok: bool,
    pub motivo: Option<String>,
    pub items: usize,
    pub duplicados: usize,
    pub nuevos: Vec<ProductoNuevo>,
    pub existentes: usize,
    pub venta_id: Option<i64>,
    pub total: f64,
}

impl ArchivoResultado {
    fn info(ok: bool, motivo: Option<String>) -> Self {
        Self {
            archivo: String::new(),
            ok,
            motivo,
            items: 0,
            duplicados: 0,
            nuevos: Vec::new(),
            existentes: 0,
            venta_id: None,
            total: 0.0,
        }
    }
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct ResumenVenta {
    pub archivo: String,
    pub venta_id: Option<i64>,
    pub items: usize,
    pub total: f64,
}

#[derive(Debug, Clone, serde::Serialize)]
pub struct TicketFallido {
    pub archivo: String,
    pub motivo: Option<String>,
}

#[derive(Debug, Default, serde::Serialize)]
pub struct EstadisticasCarpeta {
    pub total_archivos: usize,
    pub procesados: usize,
    pub exitosos: usize,
    pub errores: usize,
    pub ventas_creadas: usize,
    pub items_insertados: usize,
    pub productos_nuevos: usize,
    pub productos_existentes: usize,
    pub duplicados_detectados: usize,
    pub productos_nuevos_lista: Vec<ProductoNuevo>,
    pub resumen_ventas: Vec<ResumenVenta>,
    pub tickets_fallidos: Vec<TicketFallido>,
}

// ---------------------------------------------------------------------------
// Utilidades
// ---------------------------------------------------------------------------

fn round2(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

/// Lista de archivos .txt en la carpeta (ordenados por nombre).
pub fn obtener_archivos_txt(carpeta: &str) -> Vec<String> {
    let mut archivos: Vec<String> = Vec::new();
    if let Ok(entries) = std::fs::read_dir(carpeta) {
        for entry in entries.flatten() {
            let path = entry.path();
            if !path.is_file() {
                continue;
            }
            let es_txt = path
                .extension()
                .map(|e| e.to_string_lossy().to_ascii_lowercase() == "txt")
                .unwrap_or(false);
            if es_txt {
                archivos.push(path.to_string_lossy().to_string());
            }
        }
    }
    archivos.sort();
    archivos
}

fn nombre_de_archivo(ruta: &str) -> String {
    Path::new(ruta)
        .file_name()
        .map(|n| n.to_string_lossy().to_string())
        .unwrap_or_else(|| ruta.to_string())
}

/// Igual que Python `open(..., errors="ignore")`: bytes inválidos se descartan.
fn leer_archivo_tolerante(ruta: &str) -> std::io::Result<String> {
    let bytes = std::fs::read(ruta)?;
    Ok(String::from_utf8_lossy(&bytes).into_owned())
}

/// Suma de totales de línea. Espejo de `sum(i.get("total",0) or cant*precio)`.
fn calcular_subtotal(items: &[Item]) -> f64 {
    items
        .iter()
        .map(|i| {
            if i.total != 0.0 {
                i.total
            } else {
                i.cantidad * i.precio_unitario
            }
        })
        .sum()
}

fn extraer_cajero(texto: &str) -> String {
    for linea in texto.lines().take(10) {
        let lower = linea.to_lowercase();
        if lower.contains("cajero") || lower.contains("empleado") || lower.contains("vendedor") {
            if let Some(idx) = linea.find(':') {
                return linea[idx + 1..].trim().to_string();
            }
        }
    }
    "SISTEMA".to_string()
}

/// Precarga el set de claves `NOMBRE|precio` ya existentes en `productos`.
fn cargar_estado(db_path: &str) -> HashSet<String> {
    let mut vistos = HashSet::new();
    if let Ok(conn) = Connection::open(db_path) {
        if let Ok(mut stmt) = conn.prepare("SELECT nombre, precio_venta FROM productos") {
            if let Ok(rows) = stmt.query_map([], |row| {
                let nombre: String = row.get(0)?;
                let precio: f64 = row.get(1)?;
                Ok((nombre, precio))
            }) {
                for row in rows.flatten() {
                    let (nombre, precio) = row;
                    vistos.insert(format!("{}|{:.2}", nombre.trim().to_uppercase(), precio));
                }
            }
        }
    }
    vistos
}

fn insertar_venta(
    conn: &Connection,
    items: &[Item],
    cajero: &str,
    fecha_iso: Option<&str>,
    metodo_pago: &str,
) -> Result<i64, rusqlite::Error> {
    let subtotal = calcular_subtotal(items);
    let iva = round2(subtotal * 0.16);
    let total = round2(subtotal + iva);

    match fecha_iso {
        Some(fecha) => {
            conn.execute(
                "INSERT INTO ventas (total, subtotal, iva, cajero, metodo_pago, estado, fecha)
                 VALUES (?1, ?2, ?3, ?4, ?5, ?6, ?7)",
                params![total, subtotal, iva, cajero, metodo_pago, "completada", fecha],
            )?;
        }
        None => {
            conn.execute(
                "INSERT INTO ventas (total, subtotal, iva, cajero, metodo_pago, estado)
                 VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
                params![total, subtotal, iva, cajero, metodo_pago, "completada"],
            )?;
        }
    }
    let venta_id = conn.last_insert_rowid();

    for item in items {
        let sub = round2(item.cantidad * item.precio_unitario - item.descuento.unwrap_or(0.0));

        conn.execute(
            "INSERT INTO detalle_ventas (venta_id, producto_nombre, cantidad, precio_unitario, descuento, subtotal)
             VALUES (?1, ?2, ?3, ?4, ?5, ?6)",
            params![
                venta_id,
                item.producto,
                item.cantidad,
                item.precio_unitario,
                item.descuento.unwrap_or(0.0),
                sub
            ],
        )?;

        // Actualizar stock y vendido (case-insensitive).
        conn.execute(
            "UPDATE productos SET stock = stock - ?1 WHERE LOWER(nombre) = LOWER(?2)",
            params![item.cantidad, item.producto],
        )?;
        conn.execute(
            "UPDATE productos SET vendido = vendido + ?1 WHERE LOWER(nombre) = LOWER(?2)",
            params![item.cantidad, item.producto],
        )?;
    }

    Ok(venta_id)
}

// ---------------------------------------------------------------------------
// Núcleo: procesa cada archivo con transacción propia y emite un resultado
// por canal (equivalente al generador `_procesar_archivos` de Python).
// ---------------------------------------------------------------------------

pub fn procesar_archivos(
    archivos: &[String],
    mapeo: &MapeoColumnas,
    db_path: &str,
    tx: &Sender<ArchivoResultado>,
) {
    let mut productos_vistos = cargar_estado(db_path);

    let conn = match Connection::open(db_path) {
        Ok(c) => c,
        Err(e) => {
            for archivo in archivos {
                let mut res = ArchivoResultado::info(
                    false,
                    Some(format!("no se pudo abrir la base de datos: {e}")),
                );
                res.archivo = nombre_de_archivo(archivo);
                let _ = tx.send(res);
            }
            return;
        }
    };

    for archivo in archivos {
        let nombre_archivo = nombre_de_archivo(archivo);

        let texto = match leer_archivo_tolerante(archivo) {
            Ok(t) => t,
            Err(e) => {
                let mut res = ArchivoResultado::info(false, Some(format!("error inesperado: {e}")));
                res.archivo = nombre_archivo;
                let _ = tx.send(res);
                continue;
            }
        };

        if texto.trim().is_empty() {
            let mut res = ArchivoResultado::info(false, Some("archivo vacío".to_string()));
            res.archivo = nombre_archivo;
            let _ = tx.send(res);
            continue;
        }

        let lineas: Vec<&str> = texto
            .trim()
            .lines()
            .filter(|l| !l.trim().is_empty())
            .collect();
        if lineas.is_empty() {
            let mut res = ArchivoResultado::info(false, Some("sin líneas útiles".to_string()));
            res.archivo = nombre_archivo;
            let _ = tx.send(res);
            continue;
        }

        let total_cols = lineas
            .iter()
            .map(|l| l.split_whitespace().count())
            .max()
            .unwrap_or(0);

        let mut items: Vec<Item> = Vec::new();
        let mut seen: HashSet<String> = HashSet::new();
        let mut duplicados = 0usize;
        let mut existentes = 0usize;
        let mut nuevos: Vec<ProductoNuevo> = Vec::new();

        for linea in &lineas {
            let Some(item) = parsear_linea(linea, mapeo, total_cols) else {
                continue;
            };

            let dup_key = format!("{}|{:.2}", item.producto, item.precio_unitario);
            if seen.contains(&dup_key) {
                duplicados += 1;
                continue;
            }
            seen.insert(dup_key.clone());

            if productos_vistos.contains(&dup_key) {
                existentes += 1;
            } else {
                productos_vistos.insert(dup_key);
                nuevos.push(ProductoNuevo {
                    nombre: item.producto.clone(),
                    precio: item.precio_unitario,
                });
            }
            items.push(item);
        }

        if items.is_empty() {
            let mut res = ArchivoResultado::info(
                false,
                Some("ningún producto reconocido con el mapeo actual".to_string()),
            );
            res.archivo = nombre_archivo;
            let _ = tx.send(res);
            continue;
        }

        let (fecha, hora) = extraer_fecha_hora_regex(&texto);
        let fecha_iso = fecha.map(|f| match &hora {
            Some(h) => format!("{f} {h}:00"),
            None => format!("{f} 00:00:00"),
        });

        let cajero = extraer_cajero(&texto);
        let metodo_pago = extraer_metodo_pago(&texto);
        let total_ticket = round2(calcular_subtotal(&items) * 1.16);

        if let Err(e) = conn.execute_batch("BEGIN") {
            let mut res = ArchivoResultado::info(
                false,
                Some(format!("error al insertar en DB: no se pudo iniciar la transacción ({e})")),
            );
            res.archivo = nombre_archivo;
            res.items = items.len();
            res.duplicados = duplicados;
            res.nuevos = nuevos;
            res.existentes = existentes;
            let _ = tx.send(res);
            continue;
        }

        match insertar_venta(&conn, &items, &cajero, fecha_iso.as_deref(), &metodo_pago) {
            Ok(venta_id) => {
                let _ = conn.execute_batch("COMMIT");
                let mut res = ArchivoResultado::info(true, None);
                res.archivo = nombre_archivo;
                res.items = items.len();
                res.duplicados = duplicados;
                res.nuevos = nuevos;
                res.existentes = existentes;
                res.venta_id = Some(venta_id);
                res.total = total_ticket;
                let _ = tx.send(res);
            }
            Err(e) => {
                let _ = conn.execute_batch("ROLLBACK");
                let mut res =
                    ArchivoResultado::info(false, Some(format!("error al insertar en DB: {e}")));
                res.archivo = nombre_archivo;
                res.items = items.len();
                res.duplicados = duplicados;
                res.nuevos = nuevos;
                res.existentes = existentes;
                let _ = tx.send(res);
            }
        }
    }
}

// ---------------------------------------------------------------------------
// Modo síncrono: agrega los resultados por archivo en estadísticas totales
// ---------------------------------------------------------------------------

pub fn procesar_carpeta_impl(
    archivos: Vec<String>,
    mapeo: MapeoColumnas,
    db_path: String,
) -> EstadisticasCarpeta {
    let mut stats = EstadisticasCarpeta {
        total_archivos: archivos.len(),
        ..Default::default()
    };

    let mut nombres_nuevos_vistos: HashSet<String> = HashSet::new();
    let (tx, rx) = std::sync::mpsc::channel::<ArchivoResultado>();

    procesar_archivos(&archivos, &mapeo, &db_path, &tx);
    drop(tx);

    for res in rx {
        stats.procesados += 1;
        if res.ok {
            stats.exitosos += 1;
            stats.ventas_creadas += 1;
            stats.items_insertados += res.items;
            stats.duplicados_detectados += res.duplicados;
            stats.productos_existentes += res.existentes;
            stats.productos_nuevos += res.nuevos.len();
            for nuevo in res.nuevos {
                if nombres_nuevos_vistos.insert(nuevo.nombre.clone()) {
                    stats.productos_nuevos_lista.push(nuevo);
                }
            }
            stats.resumen_ventas.push(ResumenVenta {
                archivo: res.archivo,
                venta_id: res.venta_id,
                items: res.items,
                total: res.total,
            });
        } else {
            stats.errores += 1;
            stats.tickets_fallidos.push(TicketFallido {
                archivo: res.archivo,
                motivo: res.motivo,
            });
        }
    }

    stats.productos_nuevos_lista.truncate(100);
    stats.tickets_fallidos.truncate(500);
    stats
}

// ---------------------------------------------------------------------------
// Tests (espejo de test_lote.py, con BD SQLite temporal)
// ---------------------------------------------------------------------------

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::PathBuf;
    use std::time::{SystemTime, UNIX_EPOCH};

    const MAPEO: &str = r#"{"cantidad": 0, "producto": [1], "precio_unitario": 2, "total": 3}"#;

    const TICKET: &str = "TICKET 1\n12/05/2026\n2 TAZAS $60.00 $120.00\n1 PLATO $80.00 $80.00\nTOTAL $200.00\n";

    fn mapeo() -> MapeoColumnas {
        serde_json::from_str(MAPEO).unwrap()
    }

    fn tmp_workspace(nombre: &str) -> PathBuf {
        let nanos = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_nanos();
        let dir = std::env::temp_dir().join(format!("cerebro_lote_{}_{}", nombre, nanos));
        std::fs::create_dir_all(&dir).unwrap();
        dir
    }

    fn crear_bd(dir: &Path) -> String {
        let path = dir.join("test.db");
        let conn = Connection::open(&path).unwrap();
        conn.execute_batch(
            "CREATE TABLE productos (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                nombre TEXT NOT NULL,
                precio_venta REAL DEFAULT 0,
                stock REAL DEFAULT 0,
                vendido REAL DEFAULT 0
             );
             CREATE TABLE ventas (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                total REAL, subtotal REAL, iva REAL,
                cajero TEXT, metodo_pago TEXT, estado TEXT, fecha TEXT
             );
             CREATE TABLE detalle_ventas (
                id INTEGER PRIMARY KEY AUTOINCREMENT,
                venta_id INTEGER, producto_nombre TEXT,
                cantidad REAL, precio_unitario REAL,
                descuento REAL, subtotal REAL
             );",
        )
        .unwrap();
        drop(conn);
        path.to_string_lossy().to_string()
    }

    fn escribir(dir: &Path, nombre: &str, contenido: &str) -> String {
        let p = dir.join(nombre);
        std::fs::write(&p, contenido).unwrap();
        p.to_string_lossy().to_string()
    }

    fn contar(db: &str, tabla: &str) -> i64 {
        let conn = Connection::open(db).unwrap();
        let n: i64 = conn
            .query_row(&format!("SELECT COUNT(*) FROM {tabla}"), [], |r| r.get(0))
            .unwrap();
        n
    }

    #[test]
    fn sync_crea_ventas_y_detalles() {
        let dir = tmp_workspace("ventas");
        let db = crear_bd(&dir);
        let a = escribir(&dir, "ticket1.txt", TICKET);
        let b = escribir(&dir, "ticket2.txt", TICKET);

        let stats = procesar_carpeta_impl(vec![a, b], mapeo(), db.clone());

        assert_eq!(stats.exitosos, 2);
        assert_eq!(stats.errores, 0);
        assert_eq!(stats.ventas_creadas, 2);
        assert_eq!(stats.items_insertados, 4);
        assert_eq!(contar(&db, "ventas"), 2);
        assert_eq!(contar(&db, "detalle_ventas"), 4);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn archivo_vacio_es_error_sin_venta() {
        let dir = tmp_workspace("vacio");
        let db = crear_bd(&dir);
        let a = escribir(&dir, "vacio.txt", "\n\n");

        let stats = procesar_carpeta_impl(vec![a], mapeo(), db.clone());
        assert_eq!(stats.exitosos, 0);
        assert_eq!(stats.errores, 1);
        assert_eq!(contar(&db, "ventas"), 0);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn procesar_archivos_emite_por_canal() {
        let dir = tmp_workspace("canal");
        let db = crear_bd(&dir);
        let a = escribir(&dir, "ticket1.txt", TICKET);
        let b = escribir(&dir, "ticket2.txt", TICKET);

        let (tx, rx) = std::sync::mpsc::channel::<ArchivoResultado>();
        procesar_archivos(&[a, b], &mapeo(), &db, &tx);
        drop(tx);

        let resultados: Vec<ArchivoResultado> = rx.into_iter().collect();
        assert_eq!(resultados.len(), 2);
        assert!(resultados.iter().all(|r| r.ok));
        assert_eq!(resultados.iter().map(|r| r.items).sum::<usize>(), 4);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn linea_duplicada_no_se_descuenta_dos_veces() {
        let dir = tmp_workspace("dup");
        let db = crear_bd(&dir);
        let ticket_dup =
            "TICKET 1\n12/05/2026\n2 TAZAS $60.00 $120.00\n2 TAZAS $60.00 $120.00\nTOTAL $240.00\n";
        let a = escribir(&dir, "dup.txt", ticket_dup);

        let stats = procesar_carpeta_impl(vec![a], mapeo(), db.clone());
        assert_eq!(stats.exitosos, 1);
        assert_eq!(stats.items_insertados, 1);
        assert_eq!(stats.duplicados_detectados, 1);
        assert_eq!(contar(&db, "detalle_ventas"), 1);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn stock_se_actualiza_con_cantidad_del_item() {
        let dir = tmp_workspace("stock");
        let db = crear_bd(&dir);
        let conn = Connection::open(&db).unwrap();
        conn.execute(
            "INSERT INTO productos (nombre, stock, vendido) VALUES ('TAZAS', 100, 0)",
            [],
        )
        .unwrap();
        drop(conn);

        let a = escribir(&dir, "ticket_stock.txt", TICKET);
        procesar_carpeta_impl(vec![a], mapeo(), db.clone());

        let conn = Connection::open(&db).unwrap();
        let (stock, vendido): (f64, f64) = conn
            .query_row(
                "SELECT stock, vendido FROM productos WHERE nombre = 'TAZAS'",
                [],
                |r| Ok((r.get(0)?, r.get(1)?)),
            )
            .unwrap();
        assert_eq!(stock, 98.0);
        assert_eq!(vendido, 2.0);

        let _ = std::fs::remove_dir_all(&dir);
    }

    #[test]
    fn fallo_en_un_archivo_no_afecta_a_los_demas() {
        let dir = tmp_workspace("mixto");
        let db = crear_bd(&dir);
        let bueno = escribir(&dir, "bueno.txt", TICKET);
        let solo_cabeceras = escribir(&dir, "cabeceras.txt", "GRACIAS POR SU COMPRA\nCFDI: 4D8F2A1\n");

        let stats =
            procesar_carpeta_impl(vec![bueno, solo_cabeceras], mapeo(), db.clone());

        assert_eq!(stats.exitosos, 1);
        assert_eq!(stats.errores, 1);
        assert_eq!(stats.tickets_fallidos[0].archivo, "cabeceras.txt");
        assert_eq!(contar(&db, "ventas"), 1);
        assert_eq!(contar(&db, "detalle_ventas"), 2);

        let _ = std::fs::remove_dir_all(&dir);
    }
}